File: kata-solutions/tell-dont-ask/tell-dont-ask-python/src/domain/Order.py
from decimal import Decimal

from src.domain.OrderStatus import OrderStatus
from src.domain.OrderItem import OrderItem
from src.useCase.ApprovedOrderCannotBeRejectedError import ApprovedOrderCannotBeRejectedError
from src.useCase.RejectedOrderCannotBeApprovedError import RejectedOrderCannotBeApprovedError
from src.useCase.ShippedOrdersCannotBeChangedError import ShippedOrdersCannotBeChangedError
from src.useCase.OrderCannotBeShippedError import OrderCannotBeShippedError
from src.useCase.OrderCannotBeShippedTwiceError import OrderCannotBeShippedTwiceError

class Order(object):
    def __init__(self, order_id:int, currency:str = "EUR"):
        self._status = OrderStatus.CREATED
        self._items= []
        self._currency= "EUR"
        self._total = Decimal("0.00")
        self._tax = Decimal("0.00")
        self._id = order_id

    def get_total(self):
        return self._total

    def get_currency(self):
        return self._currency

    def get_items(self):
        return self._items

    def get_tax(self):
        return self._tax

    def get_status(self):
        return self._status

    def get_id(self):
        return self._id
      
    def add_order_item(self, item: OrderItem):
        self._items.append(item)
        self._total = self._total + item.get_taxed_amount()
        self._tax= self._tax + item.get_tax()

    def approve(self):
        if self._status is OrderStatus.SHIPPED:
            raise ShippedOrdersCannotBeChangedError()

        if self._status is OrderStatus.REJECTED:
            raise RejectedOrderCannotBeApprovedError()

        self._status = OrderStatus.APPROVED

    def reject(self):
        if self._status is OrderStatus.SHIPPED:
            raise ShippedOrdersCannotBeChangedError()
            
        if self._status is OrderStatus.APPROVED:
            raise ApprovedOrderCannotBeRejectedError()

        self._status = OrderStatus.REJECTED

    def check_shipment(self):
        if self._status is OrderStatus.CREATED or self._status is OrderStatus.REJECTED:
            raise OrderCannotBeShippedError()

        if self._status is OrderStatus.SHIPPED:
            raise OrderCannotBeShippedTwiceError()

    def shipped(self):
        self._status = OrderStatus.SHIPPED

File: kata-solutions/tell-dont-ask/tell-dont-ask-java/app/src/main/java/domain/Order.java
package domain;

import java.math.BigDecimal;
import java.util.List;
import java.util.ArrayList;
import java.util.Currency;
import java.util.Collections;

import useCase.ApprovedOrderCannotBeRejectedException;
import useCase.OrderCannotBeShippedException;
import useCase.OrderCannotBeShippedTwiceException;
import useCase.RejectedOrderCannotBeApprovedException;
import useCase.ShippedOrdersCannotBeChangedException;

import static domain.OrderStatus.CREATED;
import static domain.OrderStatus.REJECTED;
import static domain.OrderStatus.SHIPPED;

public class Order {
    private BigDecimal total;
    private Currency currency;
    private List<OrderItem> items;
    private BigDecimal tax;
    private OrderStatus status;
    private int id;

    public Order(int id) {
        this.total = new BigDecimal("0.00");
        this.currency = Currency.getInstance("EUR");
        this.items = new ArrayList<>();
        this.tax = new BigDecimal("0.00");
        this.status = OrderStatus.CREATED;
        this.id = id;
    }

    public void addOrderItem(OrderItem orderItem) {
        items.add(orderItem);
        tax = tax.add(orderItem.tax);
        total = total.add(orderItem.taxedAmount);
    }

    public BigDecimal getTotal() {
        return total;
    }

    public Currency getCurrency() {
        return currency;
    }

    public List<OrderItem> getItems() {
        return Collections.unmodifiableList(items);
    }

    public BigDecimal getTax() {
        return tax;
    }

    public OrderStatus getStatus() {
        return status;
    }

    public int getId() {
        return id;
    }

    public boolean canBeShipped() {
        if (status.equals(CREATED) || status.equals(REJECTED))
            throw new OrderCannotBeShippedException();

        if (status.equals(SHIPPED))
            throw new OrderCannotBeShippedTwiceException();

        return true;
    }

    public void ship() {
        status = OrderStatus.SHIPPED;
    }

    public void reject() {
        if (status.equals(OrderStatus.SHIPPED)) 
            throw new ShippedOrdersCannotBeChangedException();

        if (status.equals(OrderStatus.APPROVED)) 
            throw new ApprovedOrderCannotBeRejectedException();

        status = OrderStatus.REJECTED;
    }

    public void approve() {
        if (status.equals(OrderStatus.SHIPPED)) 
            throw new ShippedOrdersCannotBeChangedException();

        if (status.equals(OrderStatus.REJECTED)) 
            throw new RejectedOrderCannotBeApprovedException();

        status = OrderStatus.APPROVED;
    }
}
